### PR TITLE
Extract proxy connection logic to specialized class

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently the logic to check if a connection to a remote discovery node
exists and otherwise create a proxy connection is mixed with the
collect nodes, cluster connection lifecycle, and other
RemoteClusterConnection logic. This commit introduces a specialized
RemoteConnectionManager class which handles the open connections.
Additionally, it reworks the "round-robin" proxy logic to create the list
of potential connections at connection open/close time, opposed to each
time a connection is requested.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfile,
+                              ConnectionManager.ConnectionValidator connectionValidator,
+                              ActionListener<Void> listener) throws ConnectTransportException {
+        connectionManager.connectToNode(node, connectionProfile, connectionValidator, new ActionListener<>() {
+            @Override
+            public void onResponse(Void aVoid) {
+                try {
+                    // TODO: Propagate the connection to the listener
+                    Transport.Connection newConnection = connectionManager.getConnection(node);
+                    addConnection(newConnection);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java]: If possible, I think we should purely rely on the TransportConnectionListener events to add and remove nodes, and not manually recreate that logic. In particular, the logic as it stands here has a problem (see comment below in `removeConnection`). If need be, we can add for example extend the `TransportConnectionListener.onNodeConnected` method to also contain the `Connection` object as parameter.

Note that this also means that we need strong guarantees from the TransportConnectionListener, the order in which its methods are called (I don't think we have that today) and should also have additional tests on that.

File: server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    private volatile List<Transport.Connection> connections = Collections.emptyList();
+
+    RemoteConnectionManager(String clusterAlias, ConnectionManager connectionManager) {
+        this.clusterAlias = clusterAlias;
+        this.connectionManager = connectionManager;
+    }
+
+    public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfile,
+                              ConnectionManager.ConnectionValidator connectionValidator,
+                              ActionListener<Void> listener) throws ConnectTransportException {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java]: the listener here is never called on success?

File: server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    private synchronized void removeConnection(Transport.Connection removedConnection) {
+        int newSize = this.connections.size() - 1;
+        ArrayList<Transport.Connection> newConnections = new ArrayList<>(newSize);
+        for (Transport.Connection connection : this.connections) {
+            if (connection.equals(removedConnection) == false) {
+                newConnections.add(connection);
+            }
+        }
+        assert newConnections.size() == newSize : "Expected connection count: " + newSize + ", Found: " + newConnections.size();
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java]: We can sometimes get two close events if we have two concurrent calls to `connectToNode` for the same node which give us the same connection object. Only one of these close calls removes the connections, the other becomes a NOOP and trips this assertion.

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left some comments around the caching of connections.

**Comment by ywelsch** [comment]
There are merge conflicts here (after merging your other PR). Can you bring the PR up-to-date? I will have another look then.

**Comment by tbrooks8** [comment]
> There are merge conflicts here (after merging your other PR). Can you bring the PR up-to-date? I will have another look then.

I merged.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**